### DISCUSSION TITLE
Permission denied when running renovate locally

### DISCUSSION THREAD

**Discussion by itineric** [renovatebot/renovate#18290] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.204

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I'm running renovate locally (as a human user with command line) using docker renovate/renovate:32.208.3-slim.
The command line looks like this:
`docker run --privileged --rm -u "<my uid>:<docker gid>" -e LOG_LEVEL="debug" -e GITHUB_COM_TOKEN="xxxxx" -v /var/run/docker.sock:/var/run/docker.sock -v "/path/to/config.js:/usr/src/app/config.js" -v "``pwd``/tmp:/tmp/renovate" renovate/renovate:32.208.3-slim --token "xxxx" --endpoint=http://self-hosted-gitlab/api/v4/ --platform=gitlab --dry-run username/renovate-test`

The purpose of doing that is testing renovate on specific projects before using it on our gitlab CI.
The error i'm getting is:
`You don't have permission to create /tmp/renovate/cache/containerbase`

I think it's because the docker run for renovate_sidecar is not using my "uid" on the command line (logs provided below).

Why am i using the -u with my uid ? to get the right permissions on /var/run/docker.sock

Everything runs fine on our gitlab CI with DinD. Just trying the run it locally to configure new projects before CI integration.

### Relevant debug logs

<details><summary>Logs</summary>
```
DEBUG: Executing command (repository=username/renovate-test, branch=renovate/node-minor-patch)
       "command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/gitlab/username/renovate-test\":\"/tmp/renovate/repos/gitlab/username/renovate-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-test/dev\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.1 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --no-audit\""
DEBUG: rawExec err (repository=username/renovate-test, branch=renovate/node-minor-patch)
       "err": {
         "name": "ExecError",
         "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/gitlab/username/renovate-test\":\"/tmp/renovate/repos/gitlab/username/renovate-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-test/dev\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.1 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --no-audit\"",
         "stderr": "You don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.17.1/node-v16.17.1-linux-x64.tar.xz\nbash: npm: command not found\n",
         "stdout": "installing v2 tool node v16.17.1\n",
         "options": {
           "cwd": "/tmp/renovate/repos/gitlab/username/renovate-test/dev",
           "encoding": "utf-8",
           "env": {
             "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
             "npm_config_store": "/tmp/renovate/cache/others/pnpm",
             "HOME": "/home/ubuntu",
             "PATH": "/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
             "LC_ALL": "C.UTF-8",
             "LANG": "C.UTF-8",
             "BUILDPACK_CACHE_DIR": "/tmp/renovate/cache/containerbase",
             "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 127,
         "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/gitlab/username/renovate-test\":\"/tmp/renovate/repos/gitlab/username/renovate-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-test/dev\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.1 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --no-audit\"\nYou don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.17.1/node-v16.17.1-linux-x64.tar.xz\nbash: npm: command not found\n",
         "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/gitlab/username/renovate-test\":\"/tmp/renovate/repos/gitlab/username/renovate-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-test/dev\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.1 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --no-audit\"\nYou don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.17.1/node-v16.17.1-linux-x64.tar.xz\nbash: npm: command not found\n\n    at ChildProcess.<anonymous> (/usr/src/app/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
       }
```
</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
the renovate docker image is running as userid 1000 by default, so you need to make sure tmp has permission to write for that user.

**Reply by itineric** [reply]
tmp has permission to write for anyone, the problem is that /tmp/renovate/cache/containerbase is created by renovate itself so I cannot change its permissions

**Reply by viceice** [reply]
don't mount as `/tmp/renovate`. that makes the problem. only mount `tmp:/tmp`, because renovate tries to create the renovate folder inside tmp, which causes permission issues afterwards.

**Reply by itineric** [reply]
Just tryied it, getting exactly the same error

**Reply by viceice** [reply]
what happens when you use your host /tmp ?

**Reply by itineric** [reply]
The same error, I was first using /tmp but changed it for a local tmp folder inside my home to clean-it up faster and browse it through smb share. The tmp volume is not the issue here.

**Reply by viceice** [reply]
are you using windows and trying to use a windows directory as temp for docker? that doesn't work.

**Reply by itineric** [reply]
no, all mounts are on ext4 linux
The smb share works the other way (use Windows to look through linux drives)

**Reply by viceice** [reply]
can you validate the /tmp permissions inside renovate container:

docker run .... renovate/renovate ls -la /tmp

**Reply by itineric** [reply]
```
total 12
drwxr-xr-x 3  275 5500 4096 Oct 12 17:36 .
drwxr-xr-x 1 root root 4096 Oct 12 17:38 ..
drwxr-xr-x 4  275  139 4096 Oct 12 17:36 renovate
```

275 : is my uid
5500 : is my default group id
139 : is docker group id

There is no problem writing to tmp, as you can see, the run managed to create renovate folder and a lot of other stuff.
And every created element is owned by uid 275 and gid 139.

> ls -la /tmp/renovate/cache/containerbase
```
total 8
drwxr-xr-x 2 275 139 4096 Oct 12 17:36 .
drwxr-xr-x 5 275 139 4096 Oct 12 17:37 ..
```

I really believe that something is missing on that command that runs inside renovate docker, or maybe renovate_sidecar is missing some permission stuff.

`"command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/gitlab/username/renovate-test\":\"/tmp/renovate/repos/gitlab/username/renovate-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-test/dev\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.1 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --no-audit\""`

**Reply by viceice** [reply]
do you change the docker username or group?
then you need to map the corresponding renovate docker user option too.

because by default the docker user should run as 1000:0 (ubuntu:root)

**Reply by itineric** [reply]
I do not understand your question, whats the "renovate docker user option" ?

I also tried it with "-u 1000:139" instead of "275:139", does not work either.

**Reply by viceice** [reply]
did you read the docs?

https://docs.renovatebot.com/self-hosted-configuration/#dockeruser

**Reply by itineric** [reply]
Using this option does not change anything, still getting the same error.

**Reply by viceice** [reply]
please provide the full docker command and the config you're using. (sanitized of cause)

**Reply by itineric** [reply]
I reproduced with no config at all (renovate.json just referencing the $schema), running on a gitlab project with only a package.json in the root folder.

The command:
`docker run --rm -u "275:139" -e LOG_LEVEL="debug" -e GITHUB_COM_TOKEN="xxxx" -v /var/run/docker.sock:/var/run/docker.sock renovate/renovate:32.208.3-slim --token "xxxx" --endpoint=http://gitlab-server/api/v4/ --platform=gitlab --docker-user=275:139 --dry-run username/renovate-bot-test`

> id -u
> 275

> cut -d: -f3 < <(getent group docker)
> 139

The package.json file:
```
{
  "name": "test",
  "version": "1.0.0",
  "private": true,
  "dependencies": {
    "@angular/material": "14.2.0"
  },
  "devDependencies": {},
  "engines": {
    "npm": "8.5.0"
  }
}
```

The renovate.json file:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json"
}

```


The interesting part of the log:

```
DEBUG: rawExec err (repository=username/renovate-bot-test, branch=renovate/npm-8.x)
       "err": {
         "name": "ExecError",
         "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child --user=275:139 -v \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\":\"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.18.0 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"",
         "stderr": "mkdir: cannot create directory ‘/opt/buildpack/tools/node’: Permission denied\nmkdir: cannot create directory ‘/opt/buildpack/tools/node/16.18.0’: No such file or directory\nYou don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.18.0/node-v16.18.0-linux-x64.tar.xz\nbash: npm: command not found\n",
         "stdout": "installing v2 tool node v16.18.0\n",
         "options": {
           "cwd": "/tmp/renovate/repos/gitlab/username/renovate-bot-test",
           "encoding": "utf-8",
           "env": {
             "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
             "npm_config_store": "/tmp/renovate/cache/others/pnpm",
             "HOME": "/home/ubuntu",
             "PATH": "/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
             "LC_ALL": "C.UTF-8",
             "LANG": "C.UTF-8",
             "BUILDPACK_CACHE_DIR": "/tmp/renovate/cache/containerbase",
             "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 127,
         "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child --user=275:139 -v \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\":\"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.18.0 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nmkdir: cannot create directory ‘/opt/buildpack/tools/node’: Permission denied\nmkdir: cannot create directory ‘/opt/buildpack/tools/node/16.18.0’: No such file or directory\nYou don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.18.0/node-v16.18.0-linux-x64.tar.xz\nbash: npm: command not found\n",
         "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child --user=275:139 -v \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\":\"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/gitlab/username/renovate-bot-test\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.18.0 && install-tool npm 8.5.0 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts\"\nmkdir: cannot create directory ‘/opt/buildpack/tools/node’: Permission denied\nmkdir: cannot create directory ‘/opt/buildpack/tools/node/16.18.0’: No such file or directory\nYou don't have permission to create /tmp/renovate/cache/containerbase.\ncurl: (23) Failed writing received data to disk/application\nDownload failed: https://nodejs.org/dist/v16.18.0/node-v16.18.0-linux-x64.tar.xz\nbash: npm: command not found\n\n    at ChildProcess.<anonymous> (/usr/src/app/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
       }
```

**Reply by viceice** [reply]
you forgot to mount /tmp, renovate remounts it to the sidecar docker container to share the git repo. so it must be same path on host and container. you can change it via base path config option.

**Reply by itineric** [reply]
as it does not change anything I did not keep it on the command to make it as simple as possible
using -v `pwd`/tmp:/tmp or  -v /tmp:/tmp or --base-dir does not change the error i'm getting

**Reply by viceice** [reply]
make sure /tmp/renovate doesn't exist or has appropriate permission.

**Reply by itineric** [reply]
it does not exist and /tmp has correct permissions (as checked together yesterday)

**Reply by viceice** [reply]
you need to ensure that the container user is either 1000 or a member of root group.

**Reply by itineric** [reply]
changing 275 into 1000 does not change anything (I created a user with uid 1000 to test it already)

**Reply by viceice** [reply]
you don't need to have that user on the host. your host user doesn't matter

**Reply by itineric** [reply]
id gives me 
> uid=1000(ubuntu) gid=139 groups=139

**Reply by itineric** [reply]
I tried to isolate the reproduction from the host and used docker-compose to have a dind service aside renovate as it is done when using gitlab-ci.
I am getting the exact same error, so I really thing there is an issue somewhere.

To reproduce:
```
version: "3.1"
services:
  dind:
    image: docker:20.10.18-dind
    privileged: true
    hostname: docker
    volumes:
      - ./certs:/certs
    environment:
      DOCKER_TLS_CERTDIR: /certs
      DOCKER_CERT_PATH: /certs/client
    networks:
      - renovate-network
  renovate:
    image: renovate/renovate:32.208.3-slim
    command: --dry-run path/to/project
    volumes:
      - ./certs:/certs
    environment:
      GITHUB_COM_TOKEN: xxxx
      RENOVATE_TOKEN: xxxx
      DOCKER_DRIVER: overlay2
      DOCKER_HOST: tcp://docker:2376
      DOCKER_TLS_VERIFY: 1
      DOCKER_TLS_CERTDIR: /certs
      DOCKER_CERT_PATH: /certs/client
      RENOVATE_PLATFORM: gitlab
      RENOVATE_ENDPOINT: http://gitlab-server/api/v4/
      RENOVATE_AUTODISCOVER: "false"
      LOG_LEVEL: debug
    networks:
      - renovate-network
networks:
  renovate-network:
    name: renovate

```

**Reply by viceice** [reply]
you need to create a volume and mount it as /tmp to both container

**Reply by itineric** [reply]
ok, its working with ./tmp:/tmp volume but I do not get the difference between the working docker-compose and the non-working docker command

**Reply by viceice** [reply]
you probably mixed something up with permissions and mounts 🤷‍♂️

**Reply by itineric** [reply]
same user, same host, same folder(s)
I reproduced the problem on another server when using "docker run".
Don't you get the error ?

The only difference I see if the `-v /var/run/docker.sock:/var/run/docker.sock` vs `DOCKER_HOST: tcp://docker:2376`

**Reply by viceice** [reply]
if I follow this i don't get any errors

https://github.com/renovatebot/docker-renovate#samples


did you some special docker setup?

**Reply by itineric** [reply]
genuine docker 20.10.12
My docker run command is the sample one to which I had do add the -u instruction so that the gid reflects the docker gid of my host (otherwise I get an permission error that seems legit)
